Support USE statement

cvybhu opened the pull request:
## Description
This PR allows to perform `USE keyspace_name` queries on the database.
Like this:
```rust
session.use_keyspace("my_keyspace").await?;
```
Or during session creation:
```rust
let session: Session = SessionBuilder::new()
    .known_node("127.0.0.1:9042")
    .use_keyspace("use_ks_test")
    .build()
    .await?;
```
## Design
New keyspace to use is being sent using channels to `ClusterWorker` and then to `NodeWorker`.
Those workers do an async `select` on their work and use keyspace requests.
This ensures that no new `Nodes`/`ConnectionKeepers` are created when keyspace is changed.
Each worker keeps `used_keyspace` and later passes it when creating `Nodes`/`ConnectionKeepers`.
I used channels because they were much easier to reason about than mutexes. Mutexes could easily have race conditions when creating new `Nodes`/`ConnectionKeepers`.

`use_keyspace` succeds if at least one node responds with `Ok`.
IO Errors are ignored - If we lost connection then we will perform `use_keyspace` on reconnect later when reconnecting is added.
Any other error causes the whole thing to fail with this error.

I thought about handling `Overloaded` errors but I'm not sure if this is a good idea.
To stay correct we would have to block all requests until the node is not overloaded anymore and then send use_keyspace and unlock all requests.
This seems complicated, I think for now we can just return `Overloaded` errors to the user.

## Fixed issues
Fixes #109

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

cvybhu commented:
@psarna @piodul could you review

psarna commented:
Regarding d1484542fa9178079692a4d5ec5c7ccd0dfed4a5 - keyspace names are also really short and this copying is most likely only done very rarely, so I think it's better to drop this commit and leave the interface cleaner.

psarna commented:
Aside from that, looks generally good.

psarna commented:
Also, we should have a spin-off issue wrt. mixing `use_keyspace` and sending raw `USE KEYSPACE` queries, which can lead to weird inconsistencies - e.g. a keyspace gets changed for a session, but only for a few connections after they reconnect. We could solve this by being explicit and issuing a warning if somebody sends a raw `USE KEYSPACE` statement while also using the `set_keyspace` mechanism.

P.S. Other drivers (gocql among them I think) explicitly forbid raw `USE KEYSPACE` statements, but I'd prefer they stay, just in case somebody wants to bulk-load lots of statements from a text file.

cvybhu commented:
Keyspace name is cloned multiple times in `ClusterWorker::send_use_keyspace` - for each Node we have to make a clone which can be sent in a channel without lifetime issues. 
In the future when  `ConnectionKeeper` gets reconnecting `NodeWorker::send_use_keyspace` will probably have to do the same.  
This means that we will have one cloned keyspace name for each ConnectionKeeper - that could be a bit much.

psarna commented:
Alright, it makes sense to reduce the memory footprint for something that's very rarely accessed.

cvybhu commented:
I added a warning when a query with the text `USE KEYSPACE` is created, but I wonder If this shouldn't be an error.
If someone wants to load queries from a file and there is `USE KEYSPACE` then the following queries probably depend on using this keyspace and everything will break

psarna commented:
I'll stick with a warning for now, since it's a good compromise - the users see that what they're doing is wrong, but it's not strictly forbidden if they really want to use it that way. An alternative would be to have a config option like "strict use keyspace check", but that sounds a little too complex for such a simple matter.

By the way, the warning text is slightly incorrect - we do *support* raw `USE KEYSPACE` statements and they actually send stuff via network, so we should just say that these queries are discouraged due to possible unexpected side effects and `use_keyspace` should be used instead.

psarna commented:
Also, one more way would be to simply assume that we know better and translate this raw query into a `use_keyspace` call - along with the warning that this raw query is silently translated to a more complex operation to avoid unexpected side effects. Maybe that's the best way to handle this issue.

piodul reviewed (approved):
Some very minor nitpicks, otherwise LGTM

cvybhu commented on scylla/src/transport/session.rs:
```diff
@@ -33,7 +33,7 @@ pub struct Session {
 /// Configuration options for [`Session`].
 /// Can be created manually, but usually it's easier to use
-/// [SessionBuilder](super::session_builder::SessionBuilder)
+/// [SessionBuilder](super::session_builder::SessionBuilder)n
```
oops fixed

cvybhu commented on scylla/src/transport/connection.rs:
```diff
@@ -567,3 +593,55 @@ impl StreamIDSet {
+    // https://docs.datastax.com/en/cql-oss/3.3/cql/cql_reference/cqlCreateKeyspace.html
+    // Despite that cassandra accepts undesrcore as first character so we do too
+    // https://github.com/scylladb/scylla/blob/62551b3bd382c7c47371eb3fc38173bd0cfed44d/test/cql-pytest/test_keyspace.py#L58
+    // https://github.com/scylladb/scylla/blob/fd1dd0eac7a303ceaa9f3ff643c4848506b0c85c/thrift/thrift_validation.cc#L76
```
fixed, also fixed the typo in `undesrcore`

cvybhu commented:
I added detecting raw `USE KEYSPACE` queries and running `session::use_keyspace`.
There is still a warning that this feature is experimental, but I moved it to `Session`. 
Putting the warning in `Query::new` was a mistake - in `Connection::use_keyspace` we create a `Query` with the text `USE {}`, which means that warnings were printed even when calling `Session::use_keyspace()`

Additionally I noticed that there is a problem with case sensitivity - `Session::use_keyspace` allowed to perform only case sensitive queries which might not be what the user wanted. I can imagine there being scripts like:
```
CREATE KEYSPACE my_keyspace ...;
USE MY_KEYSPACE;
```

Now `Session::use_keyspace` takes a `bool` which decides if the name is case sensitive and depending on this quotes are added or not.

In case of raw queries we detect if there is a quote before keyspace name and set case sensitivity accordingly.

piodul commented on scylla/src/transport/session.rs:
```diff
@@ -280,6 +280,25 @@ impl Session {
+
+            let case_sensitive: bool = keyspace_name.starts_with('"');
+
+            keyspace_name = keyspace_name.trim_matches('"');
```
I wonder if, instead of parsing USE statements, we could do the following:

- Run the statement verbatim,
- Detect that we got [`Set_keyspace`](https://github.com/apache/cassandra/blob/dfdb6e3d8ebd797bbda8bbe5e47b5228a9281839/doc/native_protocol_v4.spec#L652) response,
- Run `use_keyspace()` with the name we got in the response.

WDUT @cvybhu @psarna ? This has worse latency because we have to do two round-trips, but does not need any parsing on our side.

piodul replied:
I'm also fine if we do it in a separate issue.

psarna replied:
I agree, if we manage to not parse anything driver-side, it would be very nice. The CQL protocol does not make it particularly easy, but we should take the opportunity if possible - and here we have a specific response to handle. And I also agree it can be done as a follow-up.

cvybhu replied:
This sounds like a good idea, I'm still not sure if this parsing is 100% correct.
I will create a follow-up issue for this.

cvybhu replied:
I created #168
